recommend auth code flow using signed requests (#226)
camaraproject/IdentityAndConsentManagement: documentation/CAMARA-Security-Interoperability.md

The OIDC Authorization Code Flow is defined in [OpenID Connect](https://openid.net/specs/openid-connect-core-1_0.html)

> It is RECOMMENDED that signed authentication requests be used, as specified by [OIDC](https://openid.net/specs/openid-connect-core-1_0.html#JWTRequests). The same key MAY be used for signing the authentication request as is used for client authentication.

Reviewer: ```suggestion
It is RECOMMENDED that signed authentication requests be used, as specified by [OIDC](https://openid.net/specs/openid-connect-core-1_0.html#JWTRequests). The same key MAY be used for signing the authentication request as is used for [client authentication](#client-authentication).
```

> It is RECOMMENDED that signed authentication requests be used, as specified by [OIDC](https://openid.net/specs/openid-connect-core-1_0.html#JWTRequests). The same key MAY be used for signing the authentication request as is used for [client authentication](#client-authentication). This document recommends in the [client credentials section](#client-authentication) that the `aud` value SHOULD be the URL of the Authorization Server's [Token Endpoint](https://openid.net/specs/openid-connect-core-1_0.html#TokenEndpoint). The same recommendation is given for the `aud` value of the signed request object.

Reviewer 2: Both in [OIDC](https://openid.net/specs/openid-connect-core-1_0.html#RequestObject) and [RFC 9101](https://www.rfc-editor.org/rfc/rfc9101.html#section-4), the valid value for the `aud` field should be the issuer of the authorization server. I don't see any specification that recommends the Token Endpoint for signed request object.

Even in the [CIBA](https://openid.net/specs/openid-client-initiated-backchannel-authentication-core-1_0.html#rfc.section.7.1.1) specification, it is also established that the valid value should be the issuer of the authorization server.

Reviewer 3: @garciasolero
The CAMARA specification [specifically](https://github.com/camaraproject/IdentityAndConsentManagement/blob/main/documentation/CAMARA-Security-Interoperability.md#client-authentication) makes this recommendation.

However, this applies to client authentication, and for authorisation code flow, it is only the `/token` endpoint where the client is authenticated. For signing the authorisation request, I agree that is does not necessarily follow that the `aud` should also be the URL of the token endpoint.

Reviewer 2: Yes, you are right, hence my comment. The recommendations apply depending on what type of assertion it is, and we should be aligned with what the standards say.

Reviewer 4: [OIDC Core](https://openid.net/specs/openid-connect-core-1_0.html#JWTRequests) refers to [\[RFC9101\]](https://www.rfc-editor.org/rfc/rfc9101.html#name-explicit-endpoints), under section [Explicit Endpoints](https://www.rfc-editor.org/rfc/rfc9101.html#name-explicit-endpoints), it is RECOMMENDED good practice to explicitly state the intended interaction endpoints, in this context the Authorization endpoint (authorization_endpoint)

Reviewer 4: ```suggestion
It is RECOMMENDED that signed authentication requests be used, as specified by [OIDC](https://openid.net/specs/openid-connect-core-1_0.html#JWTRequests). The same key MAY be used for signing the authentication request as is used for [client authentication](#client-authentication). 

This document recommends the value of `aud` (Audience) claim MUST be the Authorization endpoint as RECOMMENDED by [\[RFC9101\]](https://www.rfc-editor.org/rfc/rfc9101.html#name-explicit-endpoints).
```

Author: I changed the text and hopefully improved it, so that this conversation be closed as resolved.

> It is RECOMMENDED that signed authentication requests be used, as specified by [OIDC](https://openid.net/specs/openid-connect-core-1_0.html#JWTRequests). The same key MAY be used for signing the authentication request as is used for [client authentication](#client-authentication). This document recommends in the [client credentials section](#client-authentication) that the `aud` value SHOULD be the URL of the Authorization Server's [Token Endpoint](https://openid.net/specs/openid-connect-core-1_0.html#TokenEndpoint). The same recommendation is given for the `aud` value of the signed request object.

Author: Why should we use a separate standard when OIDC already defines [signed request objects](https://openid.net/specs/openid-connect-core-1_0.html#RequestObject)?

> It is RECOMMENDED that signed authentication requests be used, as specified by [OIDC](https://openid.net/specs/openid-connect-core-1_0.html#JWTRequests). The same key MAY be used for signing the authentication request as is used for [client authentication](#client-authentication). This document recommends in the [client credentials section](#client-authentication) that the `aud` value SHOULD be the URL of the Authorization Server's [Token Endpoint](https://openid.net/specs/openid-connect-core-1_0.html#TokenEndpoint). The same recommendation is given for the `aud` value of the signed request object.

Author: OIDC says:
> Note that the Request Objects defined here are compatible with those specified by [The OAuth 2.0 Authorization Framework: JWT-Secured Authorization Request (JAR)](https://openid.net/specs/openid-connect-core-1_0.html#RFC9101) [RFC9101].
That is the only reference to RFC9101.

> 

Reviewer 4: ```suggestion
The sub (Subject) claim in the signed Request Object MUST NOT be present, to prevent [cross-JWT confusion](https://www.rfc-editor.org/rfc/rfc9101.html#name-cross-jwt-confusion).

```

Author: @mhfoo https://openid.net/specs/openid-connect-core-1_0.html#RequestObject does not mention `sub`.

> It represents the request as a JWT whose Claims are the request parameters specified in [Section 3.1.2](https://openid.net/specs/openid-connect-core-1_0.html#AuthorizationEndpoint). This JWT is called a Request Object.

I would like to "resolve" this conversation by closing it.

Author: I added a security note.

Reviewer 4: LGTM

## The Scope Parameter

> Scope values determine the specific CAMARA services being requested by the Service Provider, ject to the SP being registered to use those services. The scope values must be documented in the API OAS files by all CAMARA API projects. This document does not change OIDC definitions of scope values.

Reviewer 5: ```suggestion
Scope values determine the specific CAMARA services being requested by the Service Provider, subject to the SP being registered to use those services. The scope values must be documented in the API OAS files by all CAMARA API projects. This document does not change OIDC definitions of scope values.
```

---
**NOTE**

> Scope values are an integral part of any OAuth2 and OIDC implementation. The API implementation enforces API access based on scope (if the CAMARA API project defines scopes).

Reviewer 5: ```suggestion
Scope values are an integral part of any OAuth2 and OIDC implementation. The API implementation enforces API access based on scope (if the CAMARA API subproject defines scopes).
```
